griptape/tools/google_drive/tool.py

```diff
@@ -0,0 +1,335 @@
+    
+        except HttpError as e:
+            logging.error(e)
+            return ErrorArtifact(f"error searching for file in Google Drive: {e}")
+        except MalformedError:
+            logging.error("MalformedError occurred")
+            return ErrorArtifact("error searching for file due to malformed credentials")
+
+    def _build_client(self, scopes: list[str]) -> Resource:
+        from google.oauth2 import service_account
+        from googleapiclient.discovery import build
+
+        credentials = service_account.Credentials.from_service_account_info(
+            self.service_account_credentials, scopes=scopes
+        )
+
+        return build(
+            serviceName="drive",
+            version="v3",
+            credentials=credentials.with_subject(self.owner_email)
+        )
+
+    def _path_to_file_id(self, service, path: str) -> Optional[str]:
+        parts = path.split("/")
+        current_id = "root"
+    
+        for idx, part in enumerate(parts):
+            if idx == len(parts) - 1:  # If it's the last part
+                query = f"name='{part}' and '{current_id}' in parents"  # No mime type restriction
+            else:
+                query = f"name='{part}' and '{current_id}' in parents and mimeType='application/vnd.google-apps.folder'"
+    
+            response = service.files().list(q=query).execute()
+            files = response.get("files", [])
+    
+            if not files and idx != len(parts) - 1:  # Only create folders for non-last parts
+                folder_metadata = {"name": part, "mimeType": "application/vnd.google-apps.folder", "parents": [current_id]}
+                folder = service.files().create(body=folder_metadata, fields="id").execute()
+                current_id = folder.get("id")
+            elif files:
+                current_id = files[0]["id"]
+            else:
+                return None  # File not found
+    
+        return current_id
+
+    def _save_to_drive(self, filename: str, value: any, parent_folder_id=None) -> InfoArtifact | ErrorArtifact:
+        service = self._build_client(self.UPLOAD_FILE_SCOPES)
+    
+        if isinstance(value, str):
+            value = value.encode()
+
+        parts = filename.split("/")
+        if len(parts) > 1:
+            directory = "/".join(parts[:-1])
+            parent_folder_id = self._path_to_file_id(service, directory)
+            if not parent_folder_id:
+                return ErrorArtifact(f"Could not find folder: {directory}")
+            filename = parts[-1]
+    
+        file_metadata = {"name": filename}
+        if parent_folder_id:
+            file_metadata["parents"] = [parent_folder_id]
+    
+        media = MediaIoBaseUpload(BytesIO(value), mimetype="application/octet-stream", resumable=True)
+
+        file = service.files().create(body=file_metadata, media_body=media, fields="id").execute()
+        return InfoArtifact(file)
+
+    def _list_files(self, service, query: str) -> list:
```

Comment:
Can you specify a type for the return `list`?

Reply:
Done